crates/corro-types/src/sync.rs

```diff
@@ -85,8 +85,72 @@ pub struct SyncStateV1 {
     #[speedy(default_on_eof)]
     pub last_cleared_ts: Option<Timestamp>,
 }
-
 impl SyncStateV1 {
+    pub fn contains(&self, actor_id: &ActorId, versions: &RangeInclusive<Version>) -> bool {
+        !(self.heads.get(actor_id).cloned().unwrap_or_default() < *versions.end()
+            || self
+                .need
+                .get(actor_id)
+                .cloned()
+                .unwrap_or_default()
+                .iter()
+                .any(|x| x.start() < versions.end() && versions.start() < x.end()))
+    }
+
+    pub fn merge_needs(&mut self, needs: &HashMap<ActorId, Vec<SyncNeedV1>>) {
+        for (actor, need) in needs {
+            for n in need {
+                match n {
+                    SyncNeedV1::Full { versions } => {
+                        self.merge_full_version(*actor, versions);
+                    }
+                    SyncNeedV1::Partial { version, seqs } => {
+                        let mut delete = false;
+                        self.partial_need.entry(*actor).and_modify(|e| {
+                            if let Some(need_seqs) = e.get_mut(version) {
+                                let mut seqs_set =
+                                    RangeInclusiveSet::from_iter(need_seqs.clone().into_iter());
+                                for seq in seqs {
+                                    seqs_set.remove(seq.clone())
+                                }
+                                *need_seqs = Vec::from_iter(seqs_set.clone().into_iter());
+                                if need_seqs.is_empty() {
+                                    delete = true
+                                }
+                            };
+                        });
+
+                        // we have gotten all sequence numbers, delete need
+                        if delete {
+                            if let Some(partials) = self.partial_need.get_mut(actor) {
+                                partials.remove(version);
+                            };
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    pub fn merge_full_version(&mut self, actor_id: ActorId, version: &RangeInclusive<Version>) {
+        let head = self.heads.entry(actor_id).or_default();
+        if version.end() > head {
+            // check for gaps
+            if *head + 1 < *version.start() {
+                let range = *head + 1..=*version.start() - 1;
+                self.need.entry(actor_id).or_default().push(range);
+            }
+            *head = *version.end();
+        }
+
+        self.need.entry(actor_id).and_modify(|e| {
+            let mut set = RangeInclusiveSet::from_iter(e.clone().into_iter());
+            set.remove(version.clone());
+            *e = Vec::from_iter(set.into_iter());
```

Comment:
Same as the other comment: you should be able to mutate the entry directly instead of cloning and re-creating range inclusives.